### PR TITLE
Fix false Codecov project coverage drops by uploading R and JS reports together

### LINKED ISSUE #254: incorrect codecov failures
hi @suhaani-agarwal can you please help fix this issue?
If you look at recent commits to master https://github.com/animint/animint2/commits/master/
We see that I made a bunch of commits that only affect the documentation (CONTRIBUTING.md), not the code.
Yet the codecov report says that there was a decrease in project coverage, with respect to the previous push to master.
Image

* codecov/patch above is correct, there is no change to coverage in the patch.
* however codecov/project is obviously incorrect (doc changes do not affect code coverage), so there must be something wrong with the code coverage setup. Why does it say there is a decrease? how to fix?

does it have something to do with separate upload of R and JS coverage reports?

### PR DESCRIPTION
**Pull request by Nishita-shah1**
Fixes #254

## Summary
- Codecov `project` was failing on documentation-only commits because R and JS coverage were uploaded separately. If the JS job failed, an R-only report became the baseline, so the next full R+JS upload looked like a coverage drop.
- Coverage jobs now save artifacts instead of uploading immediately. A follow-up job uploads both reports together, and only if both R and JS jobs succeed.
- CRAN checks stay in their own job. Texlive is installed only there.

## Test plan
- [ ] Confirm CI runs `R_coverage` and `JS_coverage` in parallel, then `Upload combined coverage to Codecov`.
- [ ] Confirm the combined upload job is skipped if either coverage job fails.
- [ ] Confirm CRAN still runs independently.
- [ ] On the PR, Codecov `patch` and `project` should not fail just because one coverage job failed on a previous commit.

### CODE REVIEW

**Comment by tdhock** [comment]
@ANAMASGARD please review?

File: .github/scripts/save-r-coverage.R [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+cov <- covr::package_coverage(
+  quiet = FALSE,
+  type = "tests",
+  test_files = "tests/testthat.R"
+)
+print(cov)
+
+out_file <- "r-coverage.xml"
+covr::to_cobertura(cov, filename = out_file)
```
Comment (ANAMASGARD) [inline .github/scripts/save-r-coverage.R]: @Nishita-shah1  could you preserve the previous R coverage measurement here instead of switching to Cobertura? Codecov now sees 9494 lines instead of 8933 (+561), even though this PR does not change production code, which is why codecov/project still reports -0.10%.

Please generate the Rlang JSON equivalent to the previous covr::codecov() report and save that as the artifact, while keeping the new combined upload.

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -42,44 +38,112 @@ jobs:
...
+          test -f coverage-istanbul.json
+
+      - name: Upload R and JS coverage together
+        uses: codecov/codecov-action@v5
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          files: r-coverage.xml,coverage-istanbul.json
+          disable_search: true
+          fail_ci_if_error: true
+          verbose: true
```
Comment (ANAMASGARD) [inline .github/workflows/tests.yaml:122]:
```suggestion
      - name: Verify both coverage reports exist
        run: |
          test -s r-coverage.json
          test -s coverage-istanbul.json

      - name: Upload R and JS coverage together
        uses: codecov/codecov-action@v5
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: r-coverage.json,coverage-istanbul.json
          disable_search: true
          fail_ci_if_error: true
          verbose: true
```

Comment (ANAMASGARD) [reply .github/workflows/tests.yaml:122]: After changing the R report back to the previous Rlang JSON measurement, could you update the verification and combined upload here too? Please also use test -s so empty coverage reports fail.

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -42,44 +38,112 @@ jobs:
...
+      - name: Run R tests and save coverage
+        if: matrix.test-suite == 'R_coverage'
+        run: Rscript .github/scripts/save-r-coverage.R
+
+      - name: Upload R coverage artifact
+        if: matrix.test-suite == 'R_coverage'
+        uses: actions/upload-artifact@v4
+        with:
+          name: r-coverage
+          path: r-coverage.xml
```
Comment (ANAMASGARD) [inline .github/workflows/tests.yaml]: Please also update the earlier R artifact path from r-coverage.xml to r-coverage.json

File: tests/testthat/test-renderer3-knit-print-pdf.R [hunk 4]
```diff
@@ -1,4 +1,5 @@
 acontext("render animint in pdf")
+skip_if_not(nzchar(Sys.which("pdflatex")), "pdflatex not available")
```
Comment (ANAMASGARD) [inline tests/testthat/test-renderer3-knit-print-pdf.R:2]: @Nishita-shah1  after removing Texlive from the coverage jobs, this test skips there. The CRAN job has Texlive, but build.sh uses filter="compiler", so this renderer test is not selected there either.

could we please keep this test running in at least one CI job, so the coverage test scope remains consistent with master?

**Review by ANAMASGARD — COMMENTED** [review]
@Nishita-shah1  the combined coverage upload approach looks correct for #254. I left a few inline comments 
after these changes, could you please verify one failure case where either coverage job fails and the Codecov upload is skipped, followed by a docs-only/no-code-change run showing 0.00% project coverage change? that would confirm #254 end-to-end.

File: .github/workflows/tests.yaml [hunk 5]
```diff
@@ -42,44 +38,112 @@ jobs:
...
+        run: R CMD INSTALL .
+
+      - name: git config user.name
+        run: git config --global user.name "GitHub Actions"
+
+      - name: git config user.email
+        run: git config --global user.email toby.hocking@r-project.org
+
+      - name: run CRAN build
+        run: bash build.sh
```
Comment (ANAMASGARD) [inline .github/workflows/tests.yaml:152]: could you run the PDF renderer test here, since this is the job where Texlive is installed? currently the coverage jobs skip it, and build.sh only runs filter="compiler".
```suggestion
      - name: run PDF renderer test
        run: Rscript -e 'testthat::test_file("tests/testthat/test-renderer3-knit-print-pdf.R")'

      - name: run CRAN build
        run: bash build.sh
```

File: .github/workflows/tests.yaml [hunk 6]
```diff
@@ -42,44 +38,115 @@ jobs:
...
+        run: git config --global user.name "GitHub Actions"
+
+      - name: git config user.email
+        run: git config --global user.email toby.hocking@r-project.org
+
+      - name: run PDF renderer test
+        run: Rscript -e 'testthat::test_file("tests/testthat/test-renderer3-knit-print-pdf.R")'
+
+      - name: run CRAN build
+        run: bash build.sh
```
Comment (ANAMASGARD) [inline .github/workflows/tests.yaml:152]: sorry, my earlier suggestion missed the normal test setup. test_file() sources the helpers directly, but animint2 is not loaded, so toRGB() is unavailable. could we load the same packages as tests/testthat.R before running this test?
```suggestion
      - name: run PDF renderer test
        run: Rscript -e 'library(testthat); library(animint2); library(XML); test_file("tests/testthat/test-renderer3-knit-print-pdf.R")'

      - name: run CRAN build
        run: bash build.sh
```

**Comment by ANAMASGARD** [comment]
Sir @tdhock  the PR now looks good can you please give it a last review before merge .

**Comment by tdhock** [comment]
please resolve conflicts

**Comment by Nishita-shah1** [comment]
Hi @tdhock , i have resolved conflicts

**Comment by tdhock** [comment]
please fix?
```
* checking package subdirectories ... NOTE
Problems with news in ‘NEWS.md’:
  Cannot extract version info from the following section titles:
    Changes in development (PR#339)
```

**Comment by Nishita-shah1** [comment]
Hi @tdhock, everything is fixed now. Could you please review it

File: tests/testthat/test-renderer3-knit-print-pdf.R (hunk 4 above)
Comment (tdhock) [reply tests/testthat/test-renderer3-knit-print-pdf.R]: please undo this addition, since it is unrelated to PR subject.

**Review by tdhock — CHANGES_REQUESTED** [review]

File: NEWS.md [hunk 7]
```diff
@@ -6,10 +6,9 @@
 
 - Multi-line facet strip labels: `facet_grid()` and `facet_wrap()` now render each facet variable on its own line, matching ggplot2 (issue #262).
 
-# Changes in version 2026.6.5 (PR#336)
+# Changes in version 2026.8.18 (PR#344)
 
-- `geom(showSelected.legend=FALSE)` is the canonical opt-out for legend-driven auto-injection of showSelected, including while keeping explicit showSelected variables. (Fixed #333)
-- Breaking API cleanup: PR #292 `geom(showSelected=character())` no longer opts out; use `showSelected.legend=FALSE` instead.
+- Codecov now uploads R and JavaScript coverage together, and only after both coverage jobs succeed, so a failed JS (or R) job cannot become an incomplete project-coverage baseline.
```
Comment (tdhock) [inline NEWS.md]: can you please revise this to explain what CI jobs are defined now in tests.yaml? 
is it true that it defines R_coverage and JS_coverage, and a new upload-coverage which depends on the other two?

File: .github/workflows/tests.yaml [hunk 8]
```diff
@@ -42,44 +38,115 @@ jobs:
...
+        run: R CMD INSTALL .
+
+      - name: git config user.name
+        run: git config --global user.name "GitHub Actions"
+
+      - name: git config user.email
+        run: git config --global user.email toby.hocking@r-project.org
+
+      - name: run PDF renderer test
+        run: Rscript -e 'library(testthat); library(animint2); library(XML); test_file("tests/testthat/test-renderer3-knit-print-pdf.R")'
```
Comment (tdhock) [inline .github/workflows/tests.yaml]: please delete, should be un-necessary

**Review by tdhock — CHANGES_REQUESTED** [review]

**Comment by Nishita-shah1** [comment]
hi @tdhock , done all the changes

**Review by tdhock — APPROVED** [review]
looks good thanks
